### ISSUE TITLE
[UX] Add a summary text next to the CKEditor formatting options link, showing the active format

### ISSUE THREAD

**Issue by olafgrabienski** [backdrop/backdrop-issues#3595]
Follow-up to #3516

**Original issue title:** Add summary text for the fieldset that allows selecting a text editor

In #3516 we changed the text size and appearance of the `Formatting options` link below CKEditor. In the issue, we also discussed related improvements and it was suggested to keep making these changes later, cf. https://github.com/backdrop/backdrop-issues/issues/3516#issuecomment-461993952. I've read through the issue thread again, and here are my follow-up suggestions:

1. Add a summary text behind the formatting options link, showing the active format. See https://github.com/backdrop/backdrop-issues/issues/3516#issuecomment-460044306 and https://github.com/backdrop/backdrop-issues/issues/3516#issuecomment-460287284.

2. Change terminology from "Formatting options" to "Editor details": See https://github.com/backdrop/backdrop-issues/issues/3516#issuecomment-460832014 and https://github.com/backdrop/backdrop-issues/issues/3516#issuecomment-461584101.

I'm making the second suggestion because on most of my sites, for content editors the `Formatting options` aren't options at all. Usually, editors have only the "Filtered HTML" permission, so that there is no *option* in the fieldset but only *information*. With a name like `Editor details`, the link makes also sense when there are no options.

---
PR by @klonos: https://github.com/backdrop/backdrop/pull/2550

**Comment by olafgrabienski** [comment]
**Mockup** (made with browser dev tools):

![screen-backdrop-editor-suggestion-editor-details](https://user-images.githubusercontent.com/14188307/54614220-5715bd80-4a5c-11e9-9e6d-8f4081c812a7.png)

**Comment by klonos** [comment]
> 1. Add a summary text behind the formatting options link, showing the active format ...

I like this suggestion. We use this pattern in the taxonomy term add/edit form (`/admin/structure/taxonomy/%/add`), in the "URL settings" fieldset:

<img width="797" alt="Screen Shot 2019-03-20 at 2 08 35 am" src="https://user-images.githubusercontent.com/2423362/54618033-72151b80-4ab6-11e9-810f-348d12246632.png">

>2. Change terminology from "Formatting options" to "Editor details" ...

I believe that this is also an improvement to what we have, but curious if others have any better suggestion.

**Comment by olafgrabienski** [comment]
> We use this pattern in the taxonomy term add/edit form (`/admin/structure/taxonomy/%/add`), in the "URL settings" fieldset

Interesting, wasn't aware of it!

**Comment by klonos** [comment]
Filed a PR that implements both suggestions: https://github.com/backdrop/backdrop/pull/2550

<img width="627" alt="Screen Shot 2019-03-25 at 9 55 54 pm" src="https://user-images.githubusercontent.com/2423362/54914374-e6354080-4f48-11e9-9c86-97d1151cd8ed.png">

"Pinging" @jenlampton to chime in re the label of the fieldset.

Note: please let me know if `/core/modules/filter/filter.js` is the best option to place the JS that adds the fieldset summary, or if `/core/modules/filter/js/filter.admin.js` would be more suitable, and also the position of the code.

**Comment by klonos** [comment]
...my js-fu is not strong enough to fix the animation glitch, but we already have a separate follow-up for that here: #3596

Also suggesting 1.12.6 for milestone.

**Comment by olafgrabienski** [comment]
@klonos, thanks for making the PR, it looks good in the sandbox site!

One question: Do we need the parentheses for the summary? Looks slightly better without, in my opinion, but I'm equally fine if it's with parentheses.

Btw @klonos, `.fieldset-legend span.summary` is shown only at a min-width of 450px, which could be the approach to solve https://github.com/backdrop/backdrop-issues/issues/3617#issuecomment-476172514.

**Comment by klonos** [comment]
> One question: Do we need the parentheses for the summary?

This is the default for fieldset summaries; it was not implemented in my PR. I have filed a separate issue to deal with it: #3623

> Btw @klonos, `.fieldset-legend span.summary` is shown only at a min-width of 450px, which could be the approach to solve https://github.com/backdrop/backdrop-issues/issues/3617#issuecomment-476172514.

Yes, but this is set in the [`style.css` file of the Seven theme](https://github.com/backdrop/backdrop/blob/1.x/core/themes/seven/css/style.css#L721), which means that if we change it for one fieldset, it would change for all fieldsets. Would also need a separate issue, and would be a slightly trickier problem to solve.

**Comment by olafgrabienski** [comment]
> Also suggesting 1.12.6 for milestone.

Is there a milestone policy for improvements which include string changes?

**Comment by klonos** [comment]
AFAIK, we allow string changes along with UX improvements ...often times, the string changes are the UX improvements 😄

**Comment by Graham-72** [comment]
Is the PR working as expected in the sandbox? I ask because, having set up a new text format with several filters selected, I am not seeing the editor details that I would expect. I am seeing:
![image](https://user-images.githubusercontent.com/10559274/56683476-8952b480-66c5-11e9-954d-c36676a458a3.png)

In any case, I am not sure whether a detailed list is necessary here, particularly for non-expert users if they only have the one text format they can use.

And if the detail is not shown here, I would prefer 'Formatting options' to be replaced by 'Editing options' rather than 'Editor details', and 'Editor' to be replaced by 'Text format', the latter being what the options are named on the admin/config/content/formats page. I think this would be more consistent.

**Comment by jenlampton** [comment]
> if they only have the one text format they can use.

If they only have access to one, they won't get an option to select a format at all. I don't recall if they get any description text, my bet is that they don't.

> and 'Editor' to be replaced by 'Text format', the latter being what the options are named on the admin/config/content/formats page. I think this would be more consistent.

We tried to remove the language `Text format` from Backdrop core because that phrase won't make sense to anyone who's not specifically worked with it in Drupal 7.

The kinds of people who are using this interface are the *least* likely to have ever seen that crazy back-end label before. This is a page for any old editor - not a high level admin who has access to the text formats page.

The text an editor expects to see here is something like `Editor`. I agree that we should be consistent - and that's why when we shipped with Backdrop 1.0 the back-end text also said `Editor`. (I somehow missed the opportunity to argue against that change, and was very sad when I saw the label reappear).

**Comment by Graham-72** [comment]
> Hmm... What are you expecting to see?

My preference would be not to see any details here! I was just trying to deduce the intention of the PR.

**Comment by jenlampton** [comment]
> My preference would be not to see any details here!

Me too! Maybe the PR needs a rebase & new sandbox...

**Comment by Graham-72** [comment]
> If they only have access to one editor, they won't get an option to change the editor. I don't recall if they get any description text, my bet is that they don't see that either.

I would like this to be so but my experience is that 'Formatting options' still appears. See #3705.

Or am I doing something wrong (quite possibly)?

**Comment by olafgrabienski** [comment]
@Graham-72 Thanks for chiming in! I'll try to recap the goals and the motivation of this PR, making it hopefully more clear:

(1) Add a summary text behind the 'formatting options' link, showing the active format: A small usability improvement, making it so that people don't have to expand the fieldset to get the most important information. I guess, this part of the PR isn't controversial so far.

(2) Change terminology from "Formatting options" to "Editor details": The motivation was the (common) use case where content editors have only the "Filtered HTML" permission, resulting in no 'option' in the fieldset, but information, for instance:

- `Use [fn]...[/fn] (or <fn>...</fn>) to insert automatically numbered footnotes.`
- `You may insert videos with [video:URL]`
- `Web page addresses and e-mail addresses turn into links automatically.`

In this case, a name containing the word 'details' makes more sense than 'options'.

Seeing your screenshot in https://github.com/backdrop/backdrop-issues/issues/3595#issuecomment-486372076 I guess you've tested the PR with a different use case: When an editor (or an administrator) has permission for different 'text formats'. In this case, you can choose between options, as before the PR, and the improvement (2) doesn't really matter.

**Regarding the state of the PR**, I'm a bit confused at the moment. I had a look at the sandbox a couple of weeks ago, and it seemed to look fine. Now I've had another look as a test user of role "Content editor", and in this case I don't see any information in the fieldset nor in the description, both are empty. In my opinion, this information should be there, also with the PR. @klonos Can you have a look at it?

**Comment by klonos** [comment]
>> My preference would be not to see any details here!
>
>Me too! Maybe the PR needs a rebase & new sandbox...

The only things that the PR is changing is the label of the fieldset + it also adds a summary text in the fieldset, with the name of the text format chosen. So nothing to do with the actual content of the fieldset.

Having said that, I did notice that not all details of the selected text format are shown all the times, but not only on the PR sandbox. Point is that this is a separate problem.

**Comment by jenlampton** [comment]
Maybe we should hold off on changing the label of the fieldset until we decide what to call an input format / text format / editor type / filter set / whatchamacallit?

**Comment by klonos** [comment]
Sure @jenlampton ...so retitle this issue to be just about adding the summary, and then remove the fieldset title change bit from the PR?

**Comment by klonos** [comment]
😆

**Comment by jenlampton** [comment]
Yeah, that's what I as thinking. then we can get this into 1.13 :)

**Comment by klonos** [comment]
Sure. Doing...

**Comment by klonos** [comment]
...done! And I agree that this is simple enough that can get in with 1.12.7 🙂

**Comment by klonos** [comment]
...lets leave the label bit for #3709 and perhaps separate issue (if one does not already exist) for what goes inside the fieldset.

**Comment by jenlampton** [comment]
I'm not sure this PR does what it's supposed to. Can we please get an issue summary update with the expected change, so we can more easily test the PR?

**Comment by olafgrabienski** [comment]
> I'm not sure this PR does what it's supposed to. Can we please get an issue summary update with the expected change, so we can more easily test the PR?

@jenlampton I've tried to update the issue description but wasn't able to save it. Can you give it a try? Here's what I wanted to put in the description (i.e. the first of both suggestions):

> Follow-up to #3516: Add a summary text next to the CKEditor formatting options link, showing the active format.

(then a heading "Original issue description" followed by the current description)

Also, the PR sandbox website isn't working. @klonos Could you try to restart it?

**Comment by jenlampton** [comment]
@olafgrabienski I wasn't exactly sure what changes you waned, but I changed the issue title, and added the original title into the top post. Did I do it right? Or, did you want me to change the text for Suggestion #1?

**Comment by olafgrabienski** [comment]
Thanks @jenlampton - changing the issue title is a good idea!

Actually I wanted to remove the 2nd item (*2. Change terminology from "Formatting options" ...*), or rather put it below a heading "Original issue description", but let's stick with the title change.

**Comment by klonos** [comment]
I've refreshed the PR sandbox, and it works again if you wanna give it a spin @olafgrabienski and @jenlampton 😉 

There's a couple of edge cases I want to cover, so I might update the PR soon(ish), but it works as is, and the change is pretty minimal: it adds js to grab the active/selected text format, and sticks that into a summary text in the fieldset.

If we cannot reach consensus re the label of the fieldset, then we can leave that for a follow-up issue.

**Comment by klonos** [comment]
...I would really love feedback on this:

> please let me know if `/core/modules/filter/filter.js` is the best option to place the JS that adds the fieldset summary, or if `/core/modules/filter/js/filter.admin.js` would be more suitable, and also the position of the code.

**Comment by docwilmot** [comment]
> ...I would really love feedback on this:

I think thats the right place for it. `filter.admin.js` handles the filter configuration page.
Incidentally, very cool. Left one comment

**Comment by olafgrabienski** [comment]
> I've refreshed the PR sandbox, and it works again ...

Thanks for the fresh sandbox, @klonos! I've had a look, and it works for me. Regarding the fieldset label, leaving it for a follow-up sound good.

**Comment by klonos** [comment]
Thanks for the review and comments @docwilmot and @olafgrabienski 👍 

I'm setting this back to NW, to address feedback + I noticed some weirdness in cases where there's only one format available to the user: the select drop-down is not being rendered, so the summary is not updated. I have gotten it to a point where the machine name of the filter is added in the summary in that case, but I need to get the human-readable name instead.

I'll update this ticket to let you know when it's ready for another pass of 👀

**Comment by klonos** [comment]
OK, I've updated the PR, and made sure that I have added comments, to explain what's going on. Would love another pass @olafgrabienski and @docwilmot when you can spare the time.

In the case when only a single text format is available to the current user, we are rendering it as a `'#type' => 'hidden'` input. Things would have been much simpler in-code, if we were still rendering the single text format as a select (with a single value), and were hiding it either on display, or via `#states` of something.

I have tested and confirm that this also works properly with multiple CKEditor fields in a single form 😉 

To test the single-filter case, just disable either one of the "Filtered HTML" or  "Full HTML" formats (as user1 will do).

**Comment by olafgrabienski** [comment]
Have had another look at the sandbox website, and it works as expected, also with only one format available.

(Btw, great that we are able to re-enable text formats in the meantime!)

**Comment by herbdool** [comment]
Code looks good to me

**Comment by BWPanda** [comment]
Works for me too, but I've left some suggested changes in the PR.

**Comment by klonos** [comment]
Thanks @BWPanda 🙏 ...those were some great suggestions. I've addressed the comments in the PR, and it's now ready for another round of review/testing.

@BWPanda please see my comment and screenshot in the PR @BWPanda re that `var $element = $(element);`.

**Comment by BWPanda** [comment]
I made a suggestion showing what I mean about removing that line. Once that's done then this is RTBC from me.

**Comment by klonos** [comment]
Thanks @BWPanda 🙏 ...that makes sense now. I've merged your suggestion and tested in the PR sandbox and my local 👍 ...now waiting fo tests to finish.

**Comment by klonos** [comment]
...all green 🙂

**Comment by BWPanda** [comment]
Great, thanks @klonos!

**Comment by quicksketch** [comment]
There's a missing check_plain() in the PR and a minor coding standards issue. I posted the review to https://github.com/backdrop/backdrop/pull/2550#pullrequestreview-485625906

Thanks @klonos this looks great!

**Comment by klonos** [comment]
Sorry for taking so long ...all done 🙂

**Comment by indigoxela** [comment]
Tested it in the sandbox and locally, works for me.

Here's a screenshot when using Basis for content editing:
![show-text-format](https://user-images.githubusercontent.com/842066/103137591-6febe200-46ca-11eb-929b-e5f8b1f55d9e.png)

**Comment by klonos** [comment]
> Here's a screenshot when using Basis for content editing

It seems that the fieldset legend text and the new help/indication text are not properly vertically aligned in Basis. We should fix that.

**Comment by klonos** [comment]
...OK, PR updated. Diff is this:

Before:

**Comment by indigoxela** [comment]
@klonos many thanks for updating the style, the alignment is much better now.

One thing: on very small screens the legend and summary overlap. Not sure how to deal with this and if this could be a follow-up. What do you think?

![small-screen-glitch](https://user-images.githubusercontent.com/842066/103151297-11337080-477d-11eb-8c6a-81292f0639ac.png)

**Comment by olafgrabienski** [comment]
We discussed the alignment issue long time ago -- https://github.com/backdrop/backdrop/pull/2550#issuecomment-650714362 --, and @klonos opened issue https://github.com/backdrop/backdrop-issues/issues/4460, shall we close that one? And btw, why the change of `position` from `absolute` to `relative` in the current PR instead of just a small adjustment?

I think, the overlapping issue was already discussed as well, e.g. in https://github.com/backdrop/backdrop-issues/issues/3617#issuecomment-476172514 and https://github.com/backdrop/backdrop-issues/issues/3930#issuecomment-511508775. Not sure if it's really the same issue, though.

**Comment by indigoxela** [comment]
Maybe we should wait with further fieldset adjustments - especially for Seven - until #4766 is merged? And create a follow-up for Basis, which already has its own glitches with fieldsets. In any case, let's not make it worse. :smile:

**Comment by klonos** [comment]
> Maybe we should wait with further fieldset adjustments - especially for Seven - until #4766 is merged?

I agree 👍 

> One thing: on very small screens the legend and summary overlap.

I've accidentally removed the `display: none;` rule that made it so that the summary is hidden on smaller screens (the same way as this works in Seven). I've reverted that now 👍 

> why the change of `position` from `absolute` to `relative` in the current PR instead of just a small adjustment?

@olafgrabienski I'm not a CSS expert, but I couldn't get things to work properly in RTL languages when using `position: absolute;` and trying to also use `float: left;`. Feel free to either provide pointers or code snippets here, or provide a PR against mine if you can get all that to work, and I'd be more than happy to merge it 🙂 

In the meantime, I've noticed that the text color for the summary in Basis was black, which was giving it too much "attention". I couldn't understand why Seven is picking the proper color from `system.theme.css`, while Basis doesn't, so I've "manually" forced that to `#999999`. Also happy to be pointed to a more appropriate solution for that.

**Comment by olafgrabienski** [comment]
> to work properly in RTL languages

Thanks @klonos, I see!

**Comment by klonos** [comment]
> Maybe we should wait with further fieldset adjustments - especially for Seven - until #4766 is merged?

That got merged! (thanks @BWPanda 🙏 ) ...I'll rebase and work on this soon as I can.

**Comment by jenlampton** [comment]
@klonos this looks really close, just needs a minor css change in the PR.

**Comment by klonos** [comment]
Hey @jenlampton and/or @indigoxela can you please have a look at my comment in the PR: https://github.com/backdrop/backdrop/pull/2550/files#r699194141

The code suggestion you made makes things break in Basis for me (see screenshot on that same PR comment). Glad to be provided with alternative CSS that would work better.

**Comment by klonos** [comment]
...I think I fixed it with [`b81ec61` (#2550)](https://github.com/backdrop/backdrop/pull/2550/commits/b81ec61a407476e6c493d691a623b469cf2192e1). Please review.

**Comment by indigoxela** [comment]
> ...I think I fixed it 

Yes you did. :+1: 

One minor thing I don't actually get. You add/remove new lines, but I don't understand the logic. Is there a coding standard for empty lines in CSS files, I'm not aware of?

This is what we have now (rtl/ltr):

![fieldset-ltr-rtl](https://user-images.githubusercontent.com/842066/131492637-922aa5bc-e6f9-4feb-9c2d-40bce4aa36bc.png)

Another thing: the veeery light gray on white does not comply with accessibility standards - contrast is now 2.85. It requires something about `#757575` to stop the nagging.

**Comment by klonos** [comment]
> You add/remove new lines, but I don't understand the logic. Is there a coding standard for empty lines in CSS files, I'm not aware of?

It's been a while since I've filed the PR, but I believe that:
- In the case of https://github.com/backdrop/backdrop/pull/2550/files#diff-80bb6b48b5fc030b8f0cbd56b78238528febd255dffe5783155b5ad7c003c4e4L2657 I'm just removing the extra line from the end of the file. One empty line is enough.
- In the case of https://github.com/backdrop/backdrop/pull/2550/files#diff-eafaea03bcba089bf5d270c63295585d403ccc586a8d553d6e8b4bdc1e501e89L115, I'm removing the line in order to visually group the media query with the rest of the `.fieldset-legend span.summary` selectors right above it. It also brings it inline with https://github.com/backdrop/backdrop/pull/2550/files#diff-e376e77e7a112390744d75e640cc24b43108cea9c527cfcde90290cc705d7a90L767, where the media query is not separated with a blank line from the rest of the `.fieldset-legend span.summary` group of selectors. 
- In the case of https://github.com/backdrop/backdrop/pull/2550/files#diff-e376e77e7a112390744d75e640cc24b43108cea9c527cfcde90290cc705d7a90R755 I'm adding a line, to visually separate the group of `fieldset.collapsible` selectors from the group of `.fieldset-legend` selectors that follow.

Makes any sense now?

> Another thing: the veeery light gray on white does not comply with accessibility standards - contrast is now 2.85. It requires something about `#757575` to stop the nagging.

Right ...I must have picked this from another element in the theme. `#757575` gets it to pass only WCAG AA, it will need to be something like `#595959` if we want AAA. Side by side comparison:

![image](https://user-images.githubusercontent.com/2423362/131508573-38556c98-4b0c-4ca6-9ab6-af19ab387cc9.png)

Which one should I go with? Keep in mind that we don't want to be attracting attention to these elements in the node edit form.

Speaking of which, in Seven we've removed the capitalization from the fieldset legend and summary. Should we do the same here with Basis? Before/after screenshot:

![image](https://user-images.githubusercontent.com/2423362/131509257-3ed266c2-03e5-4c41-bfc3-9a7301a01414.png)

**Comment by indigoxela** [comment]
> coding standard for empty lines in CSS files ... Makes any sense now?

Ahem, sort of :grin: 

> WCAG AA ... AAA ... Keep in mind that we don't want to be attracting attention to these elements

Sure, let's not put too much attention on this side-info. IMO AA should be enough.

> in Seven we've removed the capitalization ... Should we do the same here with Basis?

Personally, I've never been a fan of all these capitalized strings, neither in Seven nor in Basis. Don't we already have an issue for that? To me this seems a little off-topic here. Hm... nope, didn't find any open issue for that. Should we open one?

**Comment by olafgrabienski** [comment]
> didn't find any open issue for that

As far as I know, there's only an issue for Seven: https://github.com/backdrop/backdrop-issues/issues/4124

**Comment by klonos** [comment]
> ...didn't find any open issue for that. Should we open one? 

Follow-up issue: #5193

**Comment by klonos** [comment]
> Ahem, sort of 😁

Lets try again, visually this time:
- Scroll to the end of the page in https://github.com/backdrop/backdrop/edit/1.x/core/modules/filter/filter.module
  <img width=50% src="https://user-images.githubusercontent.com/2423362/131549202-9dde2d44-73d7-44b6-b4c8-fa2fc291a5d1.png">
  Then read https://docs.backdropcms.org/php-standards#indenting (emphasis on the word "single")
  > All text files should end in a single newline (`\n`)
- Open https://github.com/backdrop/backdrop/blob/1.x/core/themes/basis/css/component/fieldset.css#L46 and https://github.com/backdrop/backdrop/blob/1.x/core/themes/seven/css/style.css#L706 side by side:
  ![image](https://user-images.githubusercontent.com/2423362/131550549-6a17faee-a8e3-4196-a11d-0e5ae741ea57.png)
  Notice how in Basis the LTR and the respective RTL selectors are grouped together, separated from the previous/next set by a blank line, whereas the same set of selectors in Seven are not. Since I was working on both of these files in my PR, I tried to bring some consistency, and I thought that the grouping via blank lines in Basis made sense and made things more readable.

**Comment by klonos** [comment]
> I must have picked this from another element in the theme.

I did! ...from system.theme.css: https://github.com/backdrop/backdrop/blob/1.x/core/modules/system/css/system.theme.css#L239

> Sure, let's not put too much attention on this side-info. IMO AA should be enough.

Done 👍🏼 ...see: [`83d0c66` (#2550)](https://github.com/backdrop/backdrop/pull/2550/commits/83d0c668713005a167872e56a537c0420fbd1edb)

**Comment by jenlampton** [comment]
> in Seven we've removed the capitalization ... Should we do the same here with Basis?

I think yes. 
If we treated this one instance of captialization specially in Seven, we should do the same in Basis. I think doing it globally could be done in another issue, but since we're doing the special case for Seven here, we should add the special case for basis too :)

**Comment by klonos** [comment]
> I think yes.

@jenlampton you mean as part of this PR? ...cause I've already opened #5193 for it. I'm fine either way; just let me know 🙂

**Comment by jenlampton** [comment]
in that issue we can address all Fieldset Labels. In this PR we should only target `Formatting Options`.

**Comment by klonos** [comment]
Doing...

**Comment by klonos** [comment]
...fixed with [`e9317ff` (#2550)](https://github.com/backdrop/backdrop/pull/2550/commits/e9317ff1fbf1f29604073ab44f0a2aa82245fc3f):

![image](https://user-images.githubusercontent.com/2423362/131559222-bc8be866-3162-4e47-85b0-976409c5ed27.png)

**Comment by jenlampton** [comment]
@klonos I found one line of CSS that needs to be removed, and a few suggestions about removing the compliance comments.

**Comment by klonos** [comment]
Thanks @jenlampton ...all comments addressed. Ready for another round (hopefully final 🤞🏼 )

**Comment by jenlampton** [comment]
PR looks great, RTBC!

**Comment by klonos** [comment]
Thanks @jenlampton 🙏🏼 ...now also freshly-rebased, and with all GHA tests green! 🎉 (no code changes in the rebase commit: https://github.com/backdrop/backdrop/pull/2550/commits/08a949f08b825641b5e5b290a7994372fe2e7df6)

**Comment by quicksketch** [comment]
I merged https://github.com/backdrop/backdrop/pull/2550 into 1.x for 1.20.0. Big thanks to @klonos for leading development on this effort and @olafgrabienski for request and regular feedback! Lots of team work on this one, thanks @indigoxela, @BWPanda, @herbdool, @docwilmot, and @jenlampton for the reviews and feedback.

I don't see a strong reason to pull this into 1.19.x with 1.20.0 coming out tomorrrow, so I've only merged this into 1.x. Thanks folks!
